src/components/implementation/mem_mgr/parsec.h

```diff
@@ -0,0 +1,564 @@
+/* PARSEC library implementation */
+#ifndef PARSEC_HEADER_H
+#define PARSEC_HEADER_H
+
+#include <ck_queue.h>
+#include <ck_spinlock.h>
+
+//#define IN_LINUX
+
+#ifdef IN_LINUX
+//Linux headers
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#ifndef __USE_GNU
+#define __USE_GNU
+#endif
+#include <sched.h>
+#include <pthread.h>
+#include <assert.h>
+#include <sys/time.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#else
+// Composite header files
+#include <cos_config.h>
+#include <cos_debug.h>
+//#include <print.h>
+#define printf printc
+#endif
+
+#define ACCESS_ONCE(x) (*(volatile typeof(x) *)&(x))
+
+#define rdtscll(val) __asm__ __volatile__("rdtsc" : "=A" (val))
+
+#define cmm_barrier() __asm__ __volatile__ ("" : : : "memory")
+
+#define cos_throw(label, errno) { ret = (errno); goto label; }
+
+typedef unsigned long long quie_time_t;
+
+// for 99 percentile w/ limited memory
+//#define THRES (3000)
+
+#define MEM_SIZE (128*1024*1024)
+
+//#define QUIE_QUEUE_LIMIT (1024*8)
+
+#ifndef CACHE_LINE
+#define CACHE_LINE 64
+#endif
+
+enum {
+	PARSEC_FLAG_DEACT = 1,
+};
+
+/* void * worker(void *arg); */
+
+static inline void 
+spin_delay(unsigned long long cycles)
+{
+	unsigned long long s, e, curr;
+
+	rdtscll(s); 
+	e = s + cycles;
+
+	curr = s;
+	while (curr < e) rdtscll(curr);
+
+	return;
+}
+
+void set_prio(void);
+
+void meas_sync_start(void);
+void meas_sync_end(void);
+
+#define PACKED __attribute__((packed))
+
+#ifdef IN_LINUX
+extern int cpu_assign[41];
+
+//#define CACHE_LINE (64)
+#define CACHE_ALIGNED __attribute__((aligned(CACHE_LINE)))
+#define PAGE_SIZE (4096)
+#define PAGE_ALIGNED  __attribute__((aligned(PAGE_SIZE)))
+#define EQUIESCENCE (200)
+
+/* 
+ * Return values:
+ * 0 on failure due to contention (*target != old)
+ * 1 otherwise (*target == old -> *target = updated)
+ */
+static inline int 
+cos_cas(unsigned long *target, unsigned long old, unsigned long updated)
+{
+	char z;
+	__asm__ __volatile__("lock cmpxchgl %2, %0; setz %1"
+			     : "+m" (*target),
+			       "=a" (z)
+			     : "q"  (updated),
+			       "a"  (old)
+			     : "memory", "cc");
+	return (int)z;
+}
+
+static inline void cos_mem_fence(void)
```

Comment:
yeah, we should find a place for all of these.  This is going to make porting impossibly tedious.